scrapy/utils/gz.py

```diff
@@ -30,25 +30,25 @@ def gunzip(data):
     This is resilient to CRC checksum errors.
     """
     f = GzipFile(fileobj=BytesIO(data))
-    output = b''
+    output_list = []
     chunk = b'.'
     while chunk:
         try:
             chunk = read1(f, 8196)
-            output += chunk
+            output_list.append(chunk)
         except (IOError, EOFError, struct.error):
             # complete only if there is some data, otherwise re-raise
             # see issue 87 about catching struct.error
-            # some pages are quite small so output is '' and f.extrabuf
+            # some pages are quite small so output_list is empty and f.extrabuf
             # contains the whole page content
-            if output or getattr(f, 'extrabuf', None):
+            if output_list or getattr(f, 'extrabuf', None):
                 try:
-                    output += f.extrabuf[-f.extrasize:]
+                    output_list.append(f.extrabuf[-f.extrasize:])
```

Comment:
Have you seen a speed improvement after this change? AFAIK algoritm wasn't O(N^2) in CPython because, but O(N^2) in pypy; are you using pypy?

Reply:
Thanks!

I'm using cpython, and see very big improvement, for big files this is a matter of completing within few minutes, or stalling forever.

If you refer to cpython's realloc instruction I believe for big files realloc ends up needing to copy the string frequently,  please let me know if there is another reason why this should be O(n) in Cpython)

Reply:
Yes, I was talking about optimization which CPython has. You're right it is not O(N) because of realloc, but it is still pretty fast; I've tried a couple of benchmarks, and there was no significant difference, it was not clear which one is faster:

```
In [1]: %%time
   ...: out = ""
   ...: for x in range(10000):
   ...:     out += "." * 8000
   ...:
CPU times: user 41.6 ms, sys: 31.4 ms, total: 73 ms
Wall time: 71.8 ms
```

```
In [3]: %%time
   ...: res = []
   ...: for x in range(10000):
   ...:     res.append("." * 8000)
   ...: out = "".join(res)
   ...:
CPU times: user 37.9 ms, sys: 30.2 ms, total: 68.1 ms
Wall time: 66.2 ms
```

time vary quite a lot, but overall it seems roughly the same.

BUT it seems this optimization is not implemented for bytes in Python 3! And we missed it while porting Scrapy to Python 3. In Python 2 bytes was the same as str, and had this optimization, so the code worked fine; in Python 3 only `str` has this optimization (aka `unicode` in Python 2). So the following takes forever, unlike a version with `b"".join`:

```
%%time
out = b""
for x in range(10000):
    out += b"." * 8000
``` 

I think this a bug separate from `MAX_UNCOMPRESSED_SIZE`. What do you think about opening a separate pull request with a fix? A benefit is that it could be merged faster, as it is a clear porting bug which doesn't require much discussion.

Reply:
Nice find! 

Opened #3281